src/app/clusters/service-area-server/service-area-server.cpp

```diff
@@ -0,0 +1,1104 @@
+{
+    ChipLogProgress(Zcl, "Service Area: Instance constructor");
+    mDelegate->SetInstance(this);
+}
+
+Instance::~Instance()
+{
+    InteractionModelEngine::GetInstance()->UnregisterCommandHandler(this);
+    unregisterAttributeAccessOverride(this);
+}
+
+CHIP_ERROR Instance::Init()
+{
+    ChipLogProgress(Zcl, "Service Area: INIT");
+
+    // Check if the cluster has been selected in zap
+    VerifyOrReturnError(emberAfContainsServer(mEndpointId, Id), CHIP_ERROR_INVALID_ARGUMENT,
+                        ChipLogError(Zcl, "Service Area: The cluster with Id %lu was not enabled in zap.", long(Id)));
+
+    ReturnErrorOnFailure(InteractionModelEngine::GetInstance()->RegisterCommandHandler(this));
+
+    VerifyOrReturnError(registerAttributeAccessOverride(this), CHIP_ERROR_INCORRECT_STATE);
+
+    return mDelegate->Init();
+}
+
+//*************************************************************************
+// core functions
+
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    ChipLogDetail(Zcl, "Service Area: Reading attribute");
+
+    switch (aPath.mAttributeId)
+    {
+
+    case Attributes::SupportedLocations::Id:
+        return ReadSupportedLocations(aEncoder);
+
+    case Attributes::SupportedMaps::Id:
+        return ReadSupportedMaps(aEncoder);
+
+    case Attributes::SelectedLocations::Id:
+        return ReadSelectedLocations(aEncoder);
+
+    case Attributes::CurrentLocation::Id:
+        return aEncoder.Encode(GetCurrentLocation());
+
+    case Attributes::EstimatedEndTime::Id:
+        return aEncoder.Encode(GetEstimatedEndTime());
+
+    case Attributes::Progress::Id:
+        return ReadProgress(aEncoder);
+
+    case Attributes::FeatureMap::Id:
+        return aEncoder.Encode(mFeature.Raw());
+
+    default:
+        ChipLogProgress(Zcl, "Service Area: Read unsupported attribute %u", aPath.mAttributeId);
+    }
+
+    return CHIP_NO_ERROR;
+}
+
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    ChipLogProgress(Zcl, "%s", __func__);
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case Commands::SelectLocations::Id:
+        return CommandHandlerInterface::HandleCommand<Commands::SelectLocations::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSelectLocationsCmd(ctx, req); });
+
+    case Commands::SkipCurrentLocation::Id:
+        return CommandHandlerInterface::HandleCommand<Commands::SkipCurrentLocation::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSkipCurrentLocationCmd(ctx); });
+    }
+}
+
+//*************************************************************************
+// attribute readers
+
+CHIP_ERROR Instance::ReadSupportedLocations(AttributeValueEncoder & aEncoder)
+{
+    if (mDelegate->GetNumberOfSupportedLocations() == 0)
+    {
+        return aEncoder.EncodeNull();
```

Comment:
I don't get this.  SupportedLocations is not nullable in the spec.

Reply:
This was a mistake in the spec with is being fixed in [this PR](https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/9290/files).

I know we said that this PR is meant to reflect the 0.7 ballot, but in this case, the spec missed the case of no SupportedLocations as the spec states that this attribute must contain between 1 and 255 entries.